File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
def solution(bridge_length, weight, truck_weights):
    
    from collections import deque
    
    bridge = deque([0]*bridge_length)
    total = 0
    cnt = 0
    
    while bridge:
        
        cnt += 1
        total -= bridge.popleft()
        
        if truck_weights:
            if total + truck_weights[0] <= weight:
                t = truck_weights.pop(0)
                bridge.append(t)
                total += t
            else:
                bridge.append(0)
                
    return cnt

File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let bridge = Array(bridge_length).fill(0);
    let bridge_weight = 0;
    let time = 0;
    
    for (const truck of truck_weights) {
        while (bridge_weight - bridge[0] + truck > weight) {
            bridge_weight -= bridge.shift();
            bridge.push(0);
            time++;
        }
        bridge_weight -= bridge.shift();
        bridge.push(truck);
        bridge_weight += truck;
        time++;
    }
    
    return time + bridge_length;
}